### ISSUE TITLE
TypeReference etype not set properly during writing assembly attributes and method parameters (maybe more)...

### ISSUE THREAD

**Issue by TYoung86** [jbevain/cecil#427]
For reference;
https://github.com/TYoung86/cecil/commits/6efee375659087fb31e81334f3d223a1d8d42d08

etype was almost always ElementType.None for primitive types, causing the assembly attributes to use non-standard references to system types in the core library (in this case, even referencing the System.Private.CoreLib)

the conversion of etype to a property that looks up when not set specifically resolved it for me, but not sure if this is the right approach to resolve (thus no PR)
other changes in the fork include port to new project files and making ParameterReference non-abstract (for no effective result); they'd need to be pulled out for a PR, and they actually have no impact on the output

I can provide before and after assemblies for compare if needed, but it's a pretty clear difference

**Comment by jbevain** [comment]
I'm not quite sure where to start here.

Do you have a failing test case?
An assembly that doesn't round trip properly?

**Comment by TYoung86** [comment]
Yeah - I am at a state where the Vulkan.Binder project is starting to stabilize and work is getting calmer. I'll have some time to create a full test case.

It is a bit rambly, I must have been pretty tired when I originally wrote this...

The platform was .NET Core 1.1, reading as a base and attempting to produce a .NET Standard 1.6 targeted assembly.

I can write a demonstration test case showing an assembly being generated with a CustomAttribute that has a Class or ValueType ElementType describing one or more of it's primitive (e.g. an I4) constructor arguments. PEVerify identifies the problem.

Using module.ImportReference(typeof(int)) to select a TypeReference and using module.TypeSystem.Int32 may produce different references.

It can manifest in both saving a new assembly and in saving an assembly that was read.

The version of Cecil used was based on master from 2017/06/15, commit 9d748473214628b4852038d9ac31ece918b3f1a7.

**Comment by jbevain** [comment]
A test case would be great! Thank you!

**Comment by jbevain** [comment]
But yeah, I was debugging the other day and noticed that ElementType.None was used for «we don't know, probably a class» at different places, which isn't great. Might not be an easy fix.
